### PR TITLE
Add new transition actions: Assign new IP and Replace IP

### PR DESCRIPTION
**Pull request by szok**

### CODE REVIEW

File: src/ralph/deployment/deployment.py [hunk 1]
```diff
@@ -386,6 +386,7 @@ def _assign_hostname(instance, new_hostname, net_env=None):
         kwargs['history_kwargs'][instance.pk]['hostname'] = '{}{}'.format(
             new_hostname, ' (from {})'.format(net_env) if net_env else ''
         )
+        kwargs['shared_params']['hostnames'][instances[0].pk] = new_hostname
```
Comment (mkurek) [inline src/ralph/deployment/deployment.py]: `instances[0].pk` -> `instance.pk`

File: src/ralph/deployment/deployment.py [hunk 2]
```diff
@@ -438,6 +439,92 @@ def check_ip_from_defined_network(address):
...
+            network = Network.objects.get(
+                pk=ip_or_network['value']
+            )
+            ip = network.issue_next_free_ip()
+        logger.info('Assigning {} to {}'.format(ip, instance))
+        ethernet = Ethernet.objects.create(base_object=instance)
+        logger.info('Bounding {} to {} ethernet'.format(ip, ethernet))
+        ip.ethernet = ethernet
+
+        hostname = kwargs['shared_params']['hostnames'].get(instance.pk)
```
Comment (mkurek) [inline src/ralph/deployment/deployment.py]: what if there is no `hostnames` in `kwargs['shared_params']`?

File: src/ralph/deployment/deployment.py [hunk 3]
```diff
@@ -438,6 +439,92 @@ def check_ip_from_defined_network(address):
...
+                [network.id, network]
+            )
+    return networks
+
+
+@deployment_action(
+    verbose_name=_('Replace IP'),
+    form_fields={
+        'ipaddress': {
+            'field': ChoiceFieldWithOtherOption(
```
Comment (mkurek) [inline src/ralph/deployment/deployment.py]: it should be basic choice field (we don't allow other value here - we should choose one of current IPs of this base object)

File: src/ralph/deployment/deployment.py [hunk 4]
```diff
@@ -438,6 +439,92 @@ def check_ip_from_defined_network(address):
...
+                label=_('IP Address'),
+                auto_other_choice=False,
+            ),
+            'choices': base_object_ip_choices,
+            'exclude_from_history': True,
+        },
+        'network': {
+            'field': ChoiceFieldWithOtherOption(
```
Comment (mkurek) [inline src/ralph/deployment/deployment.py]: the same here - regular choice field instead of choice with other option

**Review by mkurek — CHANGES_REQUESTED** [review]

File: src/ralph/deployment/deployment.py [hunk 5]
```diff
@@ -438,15 +459,93 @@ def check_ip_from_defined_network(address):
...
+            ),
+            'choices': base_object_network_choices,
+            'exclude_from_history': True,
+        },
+    }
+)
+def replace_ip(cls, instances, ipaddress, network, **kwargs):
+    ip = IPAddress.objects.get(pk=ipaddress)
+    network = Network.objects.get(pk=network)
+    ip.address = str(network.get_first_free_ip())
```
Comment (mkurek) [inline src/ralph/deployment/deployment.py]: some logging will be useful here (ex. 'Replacing <OLD-IP> to <NEW-IP> for <HOSTNAME>')

**Review by mkurek — CHANGES_REQUESTED** [review]

File: src/ralph/deployment/deployment.py [hunk 6]
```diff
@@ -438,15 +459,97 @@ def check_ip_from_defined_network(address):
...
+                ip.hostname = hostname
+        except KeyError:
+            pass
+
+        ip.save()
+
+
+def base_object_ip_choices(actions, objects):
+    ipaddresses = []
+    for instance in objects:
```
Comment (mkurek) [inline src/ralph/deployment/deployment.py:495]: one more thing - we should forbid running this action (Replace IP) with more than one object.

**Review by mkurek — CHANGES_REQUESTED** [review]

File: src/ralph/deployment/deployment.py (hunk 6 above)
Comment (szok) [reply src/ralph/deployment/deployment.py:495]: Done

**Review by mkurek — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**